README.md

[![Arduino CI](https://github.com/RobTillaart/Kelvin2RGB/workflows/Arduino%20CI/badge.svg)](https://github.com/marketplace/actions/arduino_ci)
[![Arduino-lint](https://github.com/RobTillaart/Kelvin2RGB/actions/workflows/arduino-lint.yml/badge.svg)](https://github.com/RobTillaart/Kelvin2RGB/actions/workflows/arduino-lint.yml)
[![JSON check](https://github.com/RobTillaart/Kelvin2RGB/actions/workflows/jsoncheck.yml/badge.svg)](https://github.com/RobTillaart/Kelvin2RGB/actions/workflows/jsoncheck.yml)
[![GitHub issues](https://img.shields.io/github/issues/RobTillaart/Kelvin2RGB.svg)](https://github.com/RobTillaart/Kelvin2RGB/issues)

[![License: MIT](https://img.shields.io/badge/license-MIT-green.svg)](https://github.com/RobTillaart/Kelvin2RGB/blob/master/LICENSE)
[![GitHub release](https://img.shields.io/github/release/RobTillaart/Kelvin2RGB.svg?maxAge=3600)](https://github.com/RobTillaart/Kelvin2RGB/releases)
[![PlatformIO Registry](https://badges.registry.platformio.org/packages/robtillaart/library/Kelvin2RGB.svg)](https://registry.platformio.org/libraries/robtillaart/Kelvin2RGB)


# Kelvin2RGB

Arduino library for converting temperature and brightness to RGB values.


## Credentials

This library is based upon an article of Tanner Helland and a related story by Neil Bartlett

- http://www.tannerhelland.com/4435/convert-temperature-rgb-algorithm-code/
- http://www.zombieprototypes.com/?p=210
- https://en.wikipedia.org/wiki/Color_temperature#Categorizing_different_lighting

There are more approximation formulas, some claim to be better,
however these are not investigated. On request these can be added.


## Description

The library converts a temperature in Kelvin and a brightness (0..100%)
 to three numbers **red**, **green** and **blue**.
These numbers are weights can be used to correct a colour image for virtual white temperature.

There are two convert functions where the **convert_NB()** is claimed to be
the more accurate one.

With the numbers **red**, **green** and **blue** calculated one can convert images 
so they will look more like taken with candle light, sunrise or sunset etc.


**pseudo code**
```cpp
Kelvin2RGB KRGB;

KRGB.convert(1850, 100);  // sunrise light factors

for each pixel in image
{
  red   *= KRGB.red();
  green *= KRGB.green();
  blue  *= KRGB.blue();
  drawPixel();
}
```

The numbers can also be used to reduce the blue channel so it has less effect 
on "getting sleepy".

The library uses floats for the R,G and B weights to keep values as accurate as possible.
Especially with images with more than 8 bits per channel this is preferred.
That said it is also possible to use this on a 565 image or to adjust colour lookup tables.


#### Related

- https://github.com/RobTillaart/Kelvin2RGB
- https://github.com/RobTillaart/map2colour


## Interface

```cpp
#include "Kelvin2RGB.h"
```

The interface is straightforward:

- **Kelvin2RGB()** constructor.
- **void begin()** resets all internal values to 0. 
    All colours, brightness and temperature.
- **void convert_TH(float temperature, float brightness = 100)**
    temperature = 0..65500   temperature below 1000 is not well defined.
    brightness = 0..100%,
- **void convert_NB(float temperature, float brightness = 100)**
    temperature = 0..65500   temperature below 1000 is not well defined.
    brightness = 0..100%,
    Is a bit more accurate and slightly slower (few %). Read link above for more information.
- **float temperature()** returns temperature, to check the value used.
- **float brightness()** returns brightness, to check the value used.
- **float red()** returns red channel weight 0.0 .. 1.0
    note this is different from Helland / Bartlett who both use an integer value 0 .. 255
- **float green()** returns green channel weight 0.0 .. 1.0
- **float blue()** returns blue channel weight 0.0 .. 1.0
- **uint32_t setRGB(float red, float green, float blue, float brightness = 100)** sets RGB values
    red, green, blue should be in 0 .. 1.0 range. brightness should be in 0..100%, Default = 100%.
    returns a 24 bit RGB value,
- **uint32_t RGB()** returns a 24 bit RGB value, 0 .. 16777215
    more efficient than 3 floats for communication.
- **uint16_t RGB565()** returns a 16 bit RGB value, 5 bits for red, 6 for green and 5 for blue.
- **uint32_t BGR()** returns a 24 bit BGR value, 0 .. 16777215
- **uint32_t CMYK()** returns a 32 bit = 4 byte CMYK value,


#### Obsolete

- **void reset()** => replaced by **begin()**


## Operations

See examples


## Future

#### Must

- improve documentations.

#### Should

- define constants like candleLight as parameter.
- investigate other formulas.
- investigate usability for RGB led strip.

#### Could

- separate brightness per colour channel to mimic "artificial illumination"  (0.2.0 ?)
- remove begin() or reset() ?
- add examples
  - ledstrip
- use a "dirty flag" to minimize math operations.

#### Wont

- investigate **RGB_10_12_10()**
  - nowhere used (not found)


## Support

If you appreciate my libraries, you can support the development and maintenance.
Improve the quality of the libraries by providing issues and Pull Requests, or
donate through PayPal or GitHub sponsors.

Thank you,



CHANGELOG.md
# Change Log Kelvin2RGB

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).


## [0.1.7] - 2023-01-30
- update readme.md
- minor edits


## [0.1.6] - 2023-01-30
- major improvement **BGR()**
- minor improvement **CMYK()**
- extend timing example.
- update readme.md
- update GitHub actions
- update license 2023

## [0.1.5] - 2022-11-14
- Add RP2040 support to build-CI.
- Add CHANGELOG.md
- extended keywords.txt
- move all code to .cpp
- fix bug in **BGR()**
- fix unit test BGR-bug

## [0.1.4] - 2021-12-20
- update library.json
- update license
- minor edits

## [0.1.3] - 2021-11-06
- update build-CI
- add badges
- add setRGB(), CMYK(), BGR(), reset();

## [0.1.2] - 2021-06-01
- add RGB565() - 16 bit colour - output

## [0.1.1] - 2020-12-30
- ??

## [0.1.0] - 2018-01-31
- initial version




test/unit_test_001.cpp
//
//    FILE: unit_test_001.cpp
//  AUTHOR: Rob Tillaart
//    DATE: 2020-12-30
// PURPOSE: unit tests for the Kelvin2RGB library
//          https://github.com/RobTillaart/Kelvin2RGB
//          https://github.com/Arduino-CI/arduino_ci/blob/master/REFERENCE.md
//

// supported assertions
// ----------------------------
// assertEqual(expected, actual);               // a == b
// assertNotEqual(unwanted, actual);            // a != b
// assertComparativeEquivalent(expected, actual);    // abs(a - b) == 0 or (!(a > b) && !(a < b))
// assertComparativeNotEquivalent(unwanted, actual); // abs(a - b) > 0  or ((a > b) || (a < b))
// assertLess(upperBound, actual);              // a < b
// assertMore(lowerBound, actual);              // a > b
// assertLessOrEqual(upperBound, actual);       // a <= b
// assertMoreOrEqual(lowerBound, actual);       // a >= b
// assertTrue(actual);
// assertFalse(actual);
// assertNull(actual);

// // special cases for floats
// assertEqualFloat(expected, actual, epsilon);    // fabs(a - b) <= epsilon
// assertNotEqualFloat(unwanted, actual, epsilon); // fabs(a - b) >= epsilon
// assertInfinity(actual);                         // isinf(a)
// assertNotInfinity(actual);                      // !isinf(a)
// assertNAN(arg);                                 // isnan(a)
// assertNotNAN(arg);                              // !isnan(a)

#include <ArduinoUnitTests.h>


#include "Arduino.h"
#include "Kelvin2RGB.h"


unittest_setup()
{
  fprintf(stderr, "KELVIN2RGB_LIB_VERSION: %s\n", (char *) KELVIN2RGB_LIB_VERSION);
}


unittest_teardown()
{
}


unittest(test_constants)
{
  assertEqual(DLS_dark           ,     0);
  assertEqual(DLS_match          ,  1700);
  assertEqual(DLS_sodiumLamp     ,  1700);
  assertEqual(DLS_candleFlame    ,  1850);
  assertEqual(DLS_sunrise        ,  1850);
  assertEqual(DLS_sunset         ,  1850);
  assertEqual(DLS_bulb           ,  2400);
  assertEqual(DLS_bulbSoftWhite  ,  2550);
  assertEqual(DLS_LEDlamp        ,  2700);
  assertEqual(DLS_warmWhite      ,  3000);
  assertEqual(DLS_studioLight    ,  3200);
  assertEqual(DLS_studioCPlight  ,  3350);
  assertEqual(DLS_daylightHorizon,  5000);
  assertEqual(DLS_flashLight     ,  5700);
  assertEqual(DLS_xenonLight     ,  6200);
  assertEqual(DLS_dayLightBright ,  6500);
  assertEqual(DLS_normal         ,  6500);
  assertEqual(DLS_screenlow      ,  6500);
  assertEqual(DLS_screenMed      ,  8000);
  assertEqual(DLS_screenHigh     ,  9500);
  assertEqual(DLS_polewardSky0   , 15000);
  assertEqual(DLS_polewardSky1   , 19000);
  assertEqual(DLS_polewardSky2   , 23000);
  assertEqual(DLS_polewardSky3   , 27000);
}



unittest(test_constructor)
{
  Kelvin2RGB tempColor;

  tmpColor.begin();

  assertEqualFloat(0, tempColor.temperature(), 0.0001);
  assertEqualFloat(0, tempColor.brightness(), 0.0001);
  assertEqualFloat(0, tempColor.red(), 0.0001);
  assertEqualFloat(0, tempColor.green(), 0.0001);
  assertEqualFloat(0, tempColor.blue(), 0.0001);
  assertEqualFloat(0, tempColor.RGB(), 0.0001);
}


unittest(test_Tanner_Helland)
{
  Kelvin2RGB tempColor;

  tmpColor.begin();

  fprintf(stderr, "DLS_warmWhite\n");
  tempColor.convert_TH(DLS_warmWhite, 100);

  assertEqualFloat(3000, tempColor.temperature(), 0.0001);
  assertEqualFloat(100, tempColor.brightness(), 0.0001);
  assertEqualFloat(1, tempColor.red(), 0.0001);
  assertEqualFloat(0.694903, tempColor.green(), 0.0001);
  assertEqualFloat(0.431048, tempColor.blue(), 0.0001);
  assertEqual(16757102, tempColor.RGB());
}


unittest(test_Neil_Bartlett)
{
  Kelvin2RGB tempColor;

  tmpColor.begin();

  fprintf(stderr, "DLS_warmWhite\n");
  tempColor.convert_NB(DLS_warmWhite, 100);

  assertEqualFloat(3000, tempColor.temperature(), 0.0001);
  assertEqualFloat(100, tempColor.brightness(), 0.0001);
  assertEqualFloat(1, tempColor.red(), 0.0001);
  assertEqualFloat(0.707636, tempColor.green(), 0.0001);
  assertEqualFloat(0.424804, tempColor.blue(), 0.0001);
  assertEqual(16757868, tempColor.RGB());
}


unittest(test_colour_spaces)
{
  Kelvin2RGB tempColor;

  tmpColor.begin();

  fprintf(stderr, "test_colour_spaces\n");
  tempColor.setRGB(0.50, 1.00, 0.75, 90);  // brightness in %

  fprintf(stderr, "BRIGHT: %f\n", tempColor.brightness());
  fprintf(stderr, "RED   : %f\n", tempColor.red());
  fprintf(stderr, "GREEN : %f\n", tempColor.green());
  fprintf(stderr, "BLUE  : %f\n", tempColor.blue());
  fprintf(stderr, "RGB   : %d\n", tempColor.RGB());
  fprintf(stderr, "BGR   : %d\n", tempColor.BGR());
  fprintf(stderr, "CMYK  : %d\n", tempColor.CMYK());
  fprintf(stderr, "RGB565: %d\n", tempColor.RGB565());

  assertEqualFloat(90, tempColor.brightness(), 0.0001);
  assertEqualFloat(0.450, tempColor.red(),   0.0001);
  assertEqualFloat(0.900, tempColor.green(), 0.0001);
  assertEqualFloat(0.675, tempColor.blue(),  0.0001);
  assertEqual(7595692, tempColor.RGB());
  assertEqual(11331187, tempColor.BGR());
  assertEqual(2130722560, tempColor.CMYK());
  assertEqual(30517, tempColor.RGB565());
}


unittest_main()


//  -- END OF FILE ---



Kelvin2RGB.cpp
//
//    FILE: Kelvin2RGB.cpp
//  AUTHOR: Rob Tillaart
// VERSION: 0.1.7
//    DATE: 2018-01-31
// PURPOSE: Arduino library for converting temperature to RGB values
//     URL: https://github.com/RobTillaart/Kelvin2RGB


#include "Kelvin2RGB.h"


#define DIVIDE_255      (0.0039215686274509803921568627451)


Kelvin2RGB::Kelvin2RGB()
{
  begin();
}

//  empty function for now, remove?
void Kelvin2RGB::begin()
{
  _temperature = 0;
  _brightness  = 0;          //  default = darkness
  _red   = 0;
  _green = 0;
  _blue  = 0;
  _rgb   = 0;
}


void Kelvin2RGB::reset()
{
  begin();
}


////////////////////////////////////////////////////////////////
//
//  Tanner Helland formulas
//
void Kelvin2RGB::convert_TH(float temperature, float brightness)
{
  _temperature = constrain(temperature, 0, 65500);
  _brightness = constrain(brightness, 0, 100);

  _red = _green = _blue = 0;
  float t = _temperature * 0.01;

  if (t <= 66)
  {
    _red = 255;
    _green = (99.4708025861 * log(t)) - 161.1195681661;
    if (t > 19)
    {
      _blue = (138.5177312231 * log(t - 10)) - 305.0447927307;
    }
    else _blue = 0;
  }
  else
  {
    _red   = 329.698727466  * pow(t - 60, -0.1332047592);
    _green = 288.1221695283 * pow(t - 60, -0.0755148492);
    _blue  = 255;
  }

  _normalize();
}


////////////////////////////////////////////////////////////////
//
//  Neil Bartlett formulas
//
void Kelvin2RGB::convert_NB(float temperature, float brightness)
{
  _temperature = constrain(temperature, 0, 65500);
  _brightness = constrain(brightness, 0, 100);

  _red = _green = _blue = 0;
  float t = _temperature * 0.01;

  if (t <= 66)
  {
    _red = 255;
    _green = t - 2;
    _green = -155.25485562709179 - 0.44596950469579133 * _green + 104.49216199393888 * log(_green);
    _blue = 0;
    if (t > 20)
    {
      _blue = t - 10;
      _blue = -254.76935184120902 + 0.8274096064007395 * _blue + 115.67994401066147 * log(_blue);
    }
  }
  else
  {
    _red = t - 55.0;
    _red = 351.97690566805693 + 0.114206453784165 * _red - 40.25366309332127 * log(_red);
    _green = t - 50.0;
    _green = 325.4494125711974 + 0.07943456536662342 * _green - 28.0852963507957 * log(_green);
    _blue = 255;
  }

  _normalize();
}


////////////////////////////////////////////////////////////////
//
//  Other functions
//

float Kelvin2RGB::temperature()
{
  return _temperature;
}


float Kelvin2RGB::brightness()
{
  return _brightness;
}


float Kelvin2RGB::red()
{
  return _red;
}


float Kelvin2RGB::green()
{
  return _green;
}


float Kelvin2RGB::blue()
{
  return _blue;
}


uint32_t Kelvin2RGB::setRGB(float red, float green, float blue, float brightness)
{
  _brightness = brightness;
  _red   = red   * 255;
  _green = green * 255;
  _blue  = blue  * 255;
  _normalize();
  return _rgb;
}


//
//  32 bit colour  (only 3 bytes used)
//
uint32_t Kelvin2RGB::RGB()
{
  return _rgb;
}



//
//  16 bit colour - of last conversion.
//
uint16_t Kelvin2RGB::RGB565()
{
  uint16_t val = 0;
  val  = uint8_t(_red   * 32);
  val <<= 6;
  val |= uint8_t(_green * 64);
  val <<= 5;
  val |= uint8_t(_blue  * 32);
  return val;
}


uint32_t Kelvin2RGB::CMYK()
{
  float k = _red;
  if (k < _green) k = _green;
  if (k < _blue) k = _blue;

  float t1 = k;
  k = 1 - k;
  float t2 = 255.0 / (1 - k);

  uint32_t c = (t1 - _red  ) * t2;
  uint32_t m = (t1 - _green) * t2;
  uint32_t y = (t1 - _blue ) * t2;

  return (c << 24) + (m << 16) + (y << 8) + (k * 255);
}


uint32_t Kelvin2RGB::BGR()
{
  return uint8_t(255 * _blue + 0.5) * 65536UL + uint8_t(255 * _green + 0.5) * 256UL + uint8_t(255 * _red + 0.5);
  //  return round(255 * 65536UL * _blue) + round(255 * 256UL * _green) + round(255 * _red);
}


/////////////////////////////////////////////////////////////
//
//  PRIVATE
//

//  expects _red, _green, _blue in the 0..255 range.
void Kelvin2RGB::_normalize()
{
  float f = 0.01 * _brightness;

  _red   = constrain(f * _red,   0, 255);
  _green = constrain(f * _green, 0, 255);
  _blue  = constrain(f * _blue,  0, 255);

  _rgb   = round(_red) * 65536UL + round(_green) * 256UL + round(_blue);

  //  divide by 255 to get factors between 0..1
  _red   *=  DIVIDE_255;
  _green *=  DIVIDE_255;
  _blue  *=  DIVIDE_255;
}


//  -- END OF FILE --



Kelvin2RGB.h
#pragma once
//
//    FILE: Kelvin2RGB.h
//  AUTHOR: Rob Tillaart
// VERSION: 0.1.7
//    DATE: 2018-01-31
// PURPOSE: Arduino library for converting temperature to RGB values
//     URL: https://github.com/RobTillaart/Kelvin2RGB


#define KELVIN2RGB_LIB_VERSION            (F("0.1.7"))

#include "Arduino.h"


//
//  Based upon article Tanner Helland  and Neil Bartlett
//  http://www.tannerhelland.com/4435/convert-temperature-rgb-algorithm-code/
//  http://www.zombieprototypes.com/?p=210
//  https://en.wikipedia.org/wiki/Color_temperature#Categorizing_different_lighting
//
//
//  based on https://en.wikipedia.org/wiki/Color_temperature#Categorizing_different_lighting
//
//      DAY LIGHT SETTING      TEMPERATURE
//
#define DLS_dark                     0
#define DLS_match                 1700
#define DLS_sodiumLamp            1700
#define DLS_candleFlame           1850
#define DLS_sunrise               1850
#define DLS_sunset                1850
#define DLS_bulb                  2400
#define DLS_bulbSoftWhite         2550
#define DLS_LEDlamp               2700
#define DLS_warmWhite             3000
#define DLS_studioLight           3200
#define DLS_studioCPlight         3350
#define DLS_daylightHorizon       5000
#define DLS_flashLight            5700
#define DLS_xenonLight            6200
#define DLS_dayLightBright        6500
#define DLS_normal                6500
#define DLS_screenlow             6500
#define DLS_screenMed             8000
#define DLS_screenHigh            9500
#define DLS_polewardSky0         15000
#define DLS_polewardSky1         19000
#define DLS_polewardSky2         23000
#define DLS_polewardSky3         27000



class Kelvin2RGB
{
public:
  Kelvin2RGB();

  void begin();   //  empty function for now, remove?
  void reset();

  //  temperature = 0..65500   brightness = 0.0 .. 100.0%
  void convert_TH(float temperature, float brightness = 100.0);
  void convert_NB(float temperature, float brightness = 100.0);

  float    temperature();
  float    brightness();;

  //  returns 0.0 .. 1.0
  float    red();
  float    green();
  float    blue();

  //  red, green, blue should be in 0 .. 1.0 range
  //  brightness should be in 0..100%,  Default = 100%,
  //  returns RGB.
  uint32_t setRGB(float red, float green, float blue, float brightness = 100.0);

  uint32_t RGB();       //  32 bit colour  (only 3 bytes used)
  uint16_t RGB565();    //  16 bit colour

  //  Experimental 0.1.3
  uint32_t CMYK();
  uint32_t BGR();


private:
  void     _normalize();

  float    _temperature = 0;
  float    _brightness  = 0;
  float    _red   = 0;
  float    _green = 0;
  float    _blue  = 0;
  uint32_t _rgb   = 0;
};


//  -- END OF FILE --

